### PR TITLE
Add support for software PWM

### LINKED ISSUE #89: Implement software pwm output with Wiring pi
DOD

- [x] Implement the softPwm
- [x] Change the logging format to include it
- [x] modify config files

### PR DESCRIPTION
**Pull request by jonapap**
Fixes #89

### CODE REVIEW

File: src/IO/gpio/PwmOutput.h [hunk 1]
```diff
@@ -1,24 +1,26 @@
...
 
 #include <string>
 #include <spdlog/logger.h>
 
-class PwmOutput: public Output
+class PwmOutput : public Output
 {
 public:
-    PwmOutput(const std::string& name, int pin);
+    PwmOutput(std::string name, int pin, bool softPWM = false);
```
Comment (ajayyy) [inline src/IO/gpio/PwmOutput.h:15]: Is there a reason this was changed?

File: projects/HotFire/config/GpioConfig.h [hunk 2]
```diff
@@ -17,22 +17,22 @@
...
+#endif
 
     #if USE_SV02
         #define SV02_NAME "SV-02"
         #define SV02_PIN 16
         #define SV02_OPEN 1
         #define SV02_CLOSE 0
         #define SV02_EVENT_ENABLE_MASK 0b100
-    #endif
+#endif
```
Comment (ajayyy) [inline projects/HotFire/config/GpioConfig.h:35]: `#endif` are indented differently than `#if USE_SV02`

File: src/IO/gpio/Gpio.h [hunk 3]
```diff
@@ -16,19 +16,21 @@
 #include <string>
 #include <map>
 
-class Gpio : public IO
-{
+class Gpio : public IO {
```
Comment (ajayyy) [inline src/IO/gpio/Gpio.h:19]: Revert this change

File: src/IO/gpio/Gpio.h [hunk 4]
```diff
@@ -37,8 +39,7 @@ class Gpio : public IO
 	std::map<std::string, DigitalOutput> digitalOutputsMap;
 	std::map<std::string, PwmOutput> pwmOutputsMap;
 
-	struct InitFlags
-	{
+	struct InitFlags {
```
Comment (ajayyy) [inline src/IO/gpio/Gpio.h:42]: Revert

File: src/IO/gpio/PwmOutput.cpp [hunk 5]
```diff
@@ -1,35 +1,54 @@
...
+        pwmSetMode(PWM_MODE_MS);
+        pwmSetRange(256);
+        pwmSetClock(192);
+    } else {
+        int status = softPwmCreate(pinNbr, 0, 100);
+        if (status == 0) {
+            SPDLOG_LOGGER_ERROR(logger, "Error while creating software PWM output: {}", errno);
+        }
+    }
+#endif
```
Comment (ajayyy) [inline src/IO/gpio/PwmOutput.cpp:36]: Newline curly brace

File: src/IO/gpio/PwmOutput.cpp [hunk 6]
```diff
@@ -1,35 +1,54 @@
...
 }
 
-bool PwmOutput::setValue(int value) {
-    if(currentState != value) {
+bool PwmOutput::setValue(int value)
+{
+    if (currentState != value) {
```
Comment (ajayyy) [inline src/IO/gpio/PwmOutput.cpp:41]: Newline curly brace

**Review by ajayyy — CHANGES_REQUESTED** [review]

File: src/IO/gpio/PwmOutput.cpp (hunk 5 above)
Comment (jonapap) [reply src/IO/gpio/PwmOutput.cpp:36]: Wait, also for `if` statements?

**Comment by ajayyy** [comment]
https://github.com/uorocketry/rocket-code-2020/pull/106

**Pull request closed without merging**